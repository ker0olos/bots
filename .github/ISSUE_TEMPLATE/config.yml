blank_issues_enabled: true
contact_links:
  - name: Discord Support Server
    url: https://discord.gg/H69RVBxeYY
    about: Please ask any questions you have here.
