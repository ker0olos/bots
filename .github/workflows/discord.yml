name: discord

on:
  push:

jobs:
  update-slash-commands:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: denoland/setup-deno@v1
        with:
          deno-version: v1.x

      - if: ${{ github.ref == 'refs/heads/main' }}
        name: Update global slash commands on production
        env:
          APP_ID: ${{ secrets.APP_ID }}
          BOT_TOKEN: ${{ secrets.BOT_TOKEN }}
        run: deno run --allow-read --allow-env --allow-net update_commands.ts

      - if: ${{ github.ref != 'refs/heads/main' }}
        name: Update guild slash commands on test server
        env:
          APP_ID: ${{ secrets.APP_ID }}
          BOT_TOKEN: ${{ secrets.BOT_TOKEN }}
          GUILD_ID: 1046932492952805446 # test server
        run: deno run --allow-read --allow-env --allow-net update_commands.ts
