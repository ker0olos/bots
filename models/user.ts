import mongo from 'https://raw.githubusercontent.com/ker0olos/mongo/v1.0.2/mod.ts';

interface User {
  /**
   * discord user id
   */
  id: string;
  /**
   * A list of discord guild ids
   */
  guilds: string[];
}

const schema = mongo.schema<User>({
  id: {
    type: String,
    required: true,
  },
  guilds: {
    type: [String],
    default: [],
  },
});

export default mongo.model<User>('users', schema);
