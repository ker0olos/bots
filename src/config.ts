const config: {
  deploy: boolean;
  appId?: string;
  publicKey?: string;
  mongoUri?: string;
  sentry?: string;
  origin?: string;
  notice?: string;
  global?: boolean;
  gacha?: boolean;
  trading?: boolean;
  stealing?: boolean;
  synthesis?: boolean;
  shop?: boolean;
  communityPacks?: boolean;
  communityPacksMaintainerAPI?: boolean;
  communityPacksBrowseAPI?: boolean;
  combat?: boolean;
  chat?: boolean;
  //
  disableImagesProxy?: boolean;
} = {
  deploy: false,
  appId: undefined,
  publicKey: undefined,
  mongoUri: undefined,
  sentry: undefined,
  origin: undefined,
  notice: undefined,
  global: undefined,
  gacha: undefined,
  trading: undefined,
  stealing: undefined,
  synthesis: undefined,
  shop: undefined,
  communityPacks: undefined,
  communityPacksMaintainerAPI: undefined,
  communityPacksBrowseAPI: undefined,
  combat: undefined,
  chat: undefined,
  //
  disableImagesProxy: undefined,
};

export async function initConfig(): Promise<void> {
  const query = await Deno.permissions.query({ name: 'env' });

  if (query?.state === 'granted') {
    config.deploy = !!Deno.env.get('DENO_DEPLOYMENT_ID');

    config.sentry = Deno.env.get('SENTRY_DSN');

    config.appId = Deno.env.get('APP_ID');

    config.publicKey = Deno.env.get('PUBLIC_KEY');

    config.mongoUri = Deno.env.get('MONGO_URI');

    config.notice = Deno.env.get('NOTICE');

    // feature flags
    config.global = !Deno.env.has('GLOBAL') ||
      Deno.env.get('GLOBAL') === '1';

    config.gacha = !Deno.env.has('GACHA') ||
      Deno.env.get('GACHA') === '1';

    config.trading = !Deno.env.has('TRADING') ||
      Deno.env.get('TRADING') === '1';

    config.stealing = !Deno.env.has('STEALING') ||
      Deno.env.get('STEALING') === '1';

    config.synthesis = !Deno.env.has('SYNTHESIS') ||
      Deno.env.get('SYNTHESIS') === '1';

    config.shop = !Deno.env.has('SHOP') ||
      Deno.env.get('SHOP') === '1';

    config.communityPacks = !Deno.env.has('COMMUNITY_PACKS') ||
      Deno.env.get('COMMUNITY_PACKS') === '1';

    config.communityPacksMaintainerAPI =
      !Deno.env.has('COMMUNITY_PACKS_MAINTAINER_API') ||
      Deno.env.get('COMMUNITY_PACKS_MAINTAINER_API') === '1';

    config.communityPacksBrowseAPI =
      !Deno.env.has('COMMUNITY_PACKS_BROWSE_API') ||
      Deno.env.get('COMMUNITY_PACKS_BROWSE_API') === '1';

    config.combat = !Deno.env.has('COMBAT') ||
      Deno.env.get('COMBAT') === '1';

    config.chat = !Deno.env.has('CHAT') ||
      Deno.env.get('CHAT') === '1';
    //
    config.disableImagesProxy = Deno.env.get('DISABLE_IMAGES_PROXY') === '1';

    config.origin = undefined;
  }
}

export function clearConfig(): void {
  Object.keys(config).forEach((key) =>
    delete config[key as keyof typeof config]
  );
}

export default config;
