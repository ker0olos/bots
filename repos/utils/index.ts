import { randint } from '../../src/utils.ts';

import * as discord from '../../src/discord.ts';

export function roll(
  { amount }: { user: discord.User; amount: number },
  { member }: discord.Interaction<unknown>,
) {
  const rolls = [];

  const dieSize = 10;
  const minSuccess = 8;

  let successes = 0;

  for (let i = 0; i < amount; i++) {
    const roll = randint(1, dieSize);

    successes += roll >= minSuccess ? 1 : 0;

    rolls.push(roll >= minSuccess ? `__${roll}__` : `${roll}`);
  }

  const plural = successes === 1 ? 'Success' : 'Successes';

  const equation = rolls.join(' + ');

  const rolledNumber =
    `\`${amount}d${dieSize}>=${minSuccess}\` \n = [ ${equation} ] \n = **${successes}** ${plural}`;

  const message = new discord.Message().setContent(
    `<@${member!.user.id}> ${rolledNumber}`,
  );

  return message;
}
