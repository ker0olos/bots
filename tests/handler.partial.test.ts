// deno-lint-ignore-file no-explicit-any

import { assertEquals } from 'https://deno.land/std@0.177.0/testing/asserts.ts';

import {
  assertSpyCall,
  stub,
} from 'https://deno.land/std@0.177.0/testing/mock.ts';

import * as discord from '../src/discord.ts';

import utils from '../src/utils.ts';
import config from '../src/config.ts';

import { handler } from '../src/interactions.ts';

import user from '../src/user.ts';
import packs from '../src/packs.ts';

Deno.test('media suggestions', async (test) => {
  await test.step('search', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'search',
        options: [{
          name: 'title',
          value: 'title',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('anime', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'anime',
        options: [{
          name: 'title',
          value: 'title',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('manga', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'manga',
        options: [{
          name: 'title',
          value: 'title',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('collection media', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'collection',
        options: [{
          type: 1,
          name: 'media',
          options: [{
            name: 'title',
            value: 'title',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('coll media', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'coll',
        options: [{
          type: 1,
          name: 'media',
          options: [{
            name: 'title',
            value: 'title',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('mm media', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'mm',
        options: [{
          type: 1,
          name: 'media',
          options: [{
            name: 'title',
            value: 'title',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        title: {
          english: 'english title',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['media', 'title'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english title',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });
});

Deno.test('character suggestions', async (test) => {
  await test.step('character', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'character',
        options: [{
          name: 'name',
          value: 'name',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('char', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'char',
        options: [{
          name: 'name',
          value: 'name',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('im', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      data: {
        name: 'im',
        options: [{
          name: 'name',
          value: 'name',
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name'],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });
});

Deno.test('party assign character suggestions', async (test) => {
  await test.step('party assign', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      member: {
        user: {
          id: 'user_id',
        },
      },
      data: {
        name: 'party',
        options: [{
          type: 1,
          name: 'assign',
          options: [{
            name: 'name',
            value: 'name',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    const usersStub = stub(
      user,
      'allCharacters',
      () => Promise.resolve([{ id: 'packId:id' }] as any),
    );

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name', 35],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      usersStub.restore();
      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('team assign', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      member: {
        user: {
          id: 'user_id',
        },
      },
      data: {
        name: 'team',
        options: [{
          type: 1,
          name: 'assign',
          options: [{
            name: 'name',
            value: 'name',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    const usersStub = stub(
      user,
      'allCharacters',
      () => Promise.resolve([{ id: 'packId:id' }] as any),
    );

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name', 35],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      usersStub.restore();
      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });

  await test.step('p assign', async () => {
    const body = JSON.stringify({
      id: 'id',
      token: 'token',
      type: discord.InteractionType.Partial,
      guild_id: 'guild_id',
      channel_id: 'channel_id',
      member: {
        user: {
          id: 'user_id',
        },
      },
      data: {
        name: 'p',
        options: [{
          type: 1,
          name: 'assign',
          options: [{
            name: 'name',
            value: 'name',
          }],
        }],
      },
    });

    const validateStub = stub(utils, 'validateRequest', () => ({} as any));

    const signatureStub = stub(utils, 'verifySignature', ({ body }) => ({
      valid: true,
      body,
    } as any));

    const searchStub = stub(packs, 'searchMany', () =>
      Promise.resolve([{
        id: 'id',
        packId: 'packId',
        name: {
          english: 'english name',
        },
      }] as any));

    const usersStub = stub(
      user,
      'allCharacters',
      () => Promise.resolve([{ id: 'packId:id' }] as any),
    );

    config.publicKey = 'publicKey';

    try {
      const request = new Request('http://localhost:8000', {
        body,
        method: 'POST',
        headers: {
          'X-Signature-Ed25519': 'ed25519',
          'X-Signature-Timestamp': 'timestamp',
        },
      });

      const response = await handler(request);

      assertSpyCall(validateStub, 0, {
        args: [request, {
          POST: {
            headers: ['X-Signature-Ed25519', 'X-Signature-Timestamp'],
          },
        }],
      });

      assertSpyCall(signatureStub, 0, {
        args: [{
          body,
          signature: 'ed25519',
          timestamp: 'timestamp',
          publicKey: 'publicKey',
        }],
      });

      assertSpyCall(searchStub, 0, {
        args: ['characters', 'name', 35],
      });

      assertEquals(response?.ok, true);
      assertEquals(response?.redirected, false);

      assertEquals(response?.status, 200);
      assertEquals(response?.statusText, 'OK');

      const form = new FormData();

      form.append(
        'payload_json',
        JSON.stringify({
          type: 8,
          data: {
            choices: [{
              name: 'english name',
              value: 'id=packId:id',
            }],
          },
        }),
      );

      assertEquals(await response?.formData(), form);
    } finally {
      delete config.publicKey;

      usersStub.restore();
      searchStub.restore();
      validateStub.restore();
      signatureStub.restore();
    }
  });
});
