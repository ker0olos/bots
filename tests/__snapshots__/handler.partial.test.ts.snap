export const snapshot = {};

snapshot[`skills > acquire 1`] = `
{
  "type": 8,
  "data": {
    "choices": [
      {
        "name": "Critical Hit",
        "value": "~crit"
      },
      {
        "name": "Enraged",
        "value": "~enrage"
      },
      {
        "name": "Stun",
        "value": "~stun"
      },
      {
        "name": "Chain Attack",
        "value": "~chain"
      },
      {
        "name": "Lifesteal",
        "value": "~lifesteal"
      },
      {
        "name": "Speed Boost",
        "value": "~speed-boost"
      },
      {
        "name": "Sneak Attack",
        "value": "~sneak"
      },
      {
        "name": "Emergency Healing",
        "value": "~heal"
      },
      {
        "name": "Defense Boost",
        "value": "~defense-boost"
      },
      {
        "name": "Slow Debuff",
        "value": "~slow-debuff"
      }
    ]
  }
}
`;

snapshot[`skills > upgrade 1`] = `
{
  "type": 8,
  "data": {
    "choices": [
      {
        "name": "Critical Hit",
        "value": "~crit"
      },
      {
        "name": "Enraged",
        "value": "~enrage"
      },
      {
        "name": "Stun",
        "value": "~stun"
      },
      {
        "name": "Chain Attack",
        "value": "~chain"
      },
      {
        "name": "Lifesteal",
        "value": "~lifesteal"
      },
      {
        "name": "Speed Boost",
        "value": "~speed-boost"
      },
      {
        "name": "Sneak Attack",
        "value": "~sneak"
      },
      {
        "name": "Emergency Healing",
        "value": "~heal"
      },
      {
        "name": "Defense Boost",
        "value": "~defense-boost"
      },
      {
        "name": "Slow Debuff",
        "value": "~slow-debuff"
      }
    ]
  }
}
`;
